Add "Last Played" column to library

ywwg opened the pull request:
https://bugs.launchpad.net/mixxx/+bug/1490818

This adds an artificial column (the product of a simple TEMPORARY VIEW query) that reports the last time a track was played according to the history playlists.

This mostly works, except if I delete an old history playlist sometimes the library view is slow to show the updated values.

ywwg committed:
last-played-datetime: initial work to create a sortable column for last played time for every track.

https://bugs.launchpad.net/mixxx/+bug/1490818

This initial version doesn't respond to changes in the playlist,
but it does do the initial population at startup.

ywwg committed:
last played time: Respond to setlog changes

ywwg committed:
last played: cleanup before review

ywwg committed:
last played: tiny optimization

ywwg commented:
Also, bindValue isn't work for some reason.  I'll try it again but it keeps saying "parameter count mismatch" even though everything looks right.

Be-ing commented on src/library/lastplayedcache.h:
```diff
@@ -0,0 +1,35 @@
+// LastPlayedCache is a helper object for creating and maintaining
+// the "last_played" temporary view on the database. It creates the
+// view and also watches for changes to the Setlog Playlists and updates
+// tracks in the cache as necessary.
```
Thanks for including this documentation :)

Be-ing replied:
Use `///` so the comment is picked up by IDEs and Doxygen

ywwg committed:
last played: triple slashes for docs

ywwg committed:
last played: this change breaks the PR.  Why doesn't bindValue work??

ywwg commented:
pushed a change that tries to use bindValue but it fails.  Can't figure out why, but the prepare statement seems to be returning false

ywwg committed:
last played: fix bindvalue uses

VIEWs can't use params.

ywwg commented:
ok bindvalue works now

Be-ing commented:
Build fails:
```
/home/be/sw/mixxx/src/library/lastplayedcache.cpp:70:9: error: non-void function 'fetchLastPlayedTime' should return a value [-Wreturn-type]
        return;
        ^
1 error generated.
```

Be-ing commented on src/track/track.cpp:
```diff
@@ -368,6 +368,20 @@ void Track::setDateAdded(const QDateTime& dateAdded) {
+
+void Track::setLastPlayedDate(const QDateTime& playedDate) {
+    QMutexLocker lock(&m_qMutex);
+    if (playedDate != m_lastPlayed) {
```
Should this be `==`? If I understand correctly, the purpose is to avoid marking the Track as dirty for a no-op, right?

ywwg replied:
lol yup

ywwg committed:
last played: oops can't return nothing

ywwg committed:
last played: oops use the right conditional

Be-ing commented:
It works, cool! I'll let others review the implementation details.

uklotzde commented on src/library/lastplayedcache.cpp:
```diff
@@ -0,0 +1,74 @@
+
+// static, so that trackdao.cpp can fetch values for initial population in the cache.
+QDateTime LastPlayedCache::fetchLastPlayedTime(const QSqlDatabase& db, TrackPointer pTrack) {
+    QSqlQuery updateQuery(db);
```
The name is misleading, this is not an update.

uklotzde replied:
Could we prepare the query once and store it as a member? We don't need to prepare the same query over and over again. This only works for the same database connection, i.e. the method would need to be non-static.

ywwg replied:
I would like it to be non-static, but trackdao doesn't have access to the lastplayedcache object created by mixxxlibraryfeature.  I could make lastplayedcache something that can be instantiated multiple times but only if there's a nice way to make sure the setup isn't done twice.  For the DB view, I can add IF NOT EXISTS, but for the qt connections I'm not sure how best to do that.   (reference counting?)

ywwg replied:
Also, trackdao has access to a database connection, but not a trackcollection

ywwg replied:
(I could just duplicate the query in trackdao but the idea was to have all the lastplayedcache functions in one file instead of spreading the logic all over mixxx)

ywwg replied:
fixed

uklotzde commented on src/library/lastplayedcache.cpp:
```diff
@@ -0,0 +1,74 @@
+    if (!updateQuery.exec()) {
+        LOG_FAILED_QUERY(updateQuery);
+    }
+    const int col = updateQuery.record().indexOf("datetime_played");
```
Looking up the column by name is not needed when you know the ordering of columns. Especially when selecting only a single column.

ywwg replied:
done

uklotzde commented on src/library/lastplayedcache.h:
```diff
@@ -0,0 +1,35 @@
+
+class TrackCollection;
+
+const QString LASTPLAYEDTABLE_NAME = "last_played";
```
This will create multiple instances of this constant for everyone who includes the header file. For a singular instance declare this constant as `extern` and define it only one in the .cpp file.

ywwg replied:
this pattern is all over mixxx fyi, see trackschema.h, defs_vinylcontrol.h, etc.  because I have a #pragma once, I don't think it'll get redefined, right?  Anyway, fixed

uklotzde commented on src/library/lastplayedcache.h:
```diff
@@ -0,0 +1,35 @@
+    Q_OBJECT
+  public:
+    LastPlayedCache(TrackCollection* trackCollection);
+    ~LastPlayedCache() {
```
`override = default;`

ywwg replied:
done

uklotzde commented on src/library/lastplayedcache.h:
```diff
@@ -0,0 +1,35 @@
+class LastPlayedCache : public QObject {
+    Q_OBJECT
+  public:
+    LastPlayedCache(TrackCollection* trackCollection);
```
`explicit`

ywwg replied:
done

uklotzde commented on src/library/lastplayedcache.h:
```diff
@@ -0,0 +1,35 @@
+    ~LastPlayedCache() {
+    }
+
+    static QDateTime fetchLastPlayedTime(const QSqlDatabase& db, TrackPointer pTrack);
```
m_pTrackCollection is a member and already holds a database connection. Why is this a static method that is passed a database connection? Do we need this flexibility?

ywwg replied:
see above -- because trackdao can't get to this object in any way I can see

ywwg replied:
fixed, I split up the object

uklotzde commented on src/library/mixxxlibraryfeature.cpp:
```diff
@@ -56,6 +57,7 @@ MixxxLibraryFeature::MixxxLibraryFeature(Library* pLibrary,
             << "library." + LIBRARYTABLE_REPLAYGAIN
             << "library." + LIBRARYTABLE_FILETYPE
             << "library." + LIBRARYTABLE_DATETIMEADDED
+            << "last_played.datetime_played"
```
The table name is defined as a constant, but you don't use it?

ywwg replied:
fixed

uklotzde commented on src/track/track.cpp:
```diff
@@ -368,6 +368,20 @@ void Track::setDateAdded(const QDateTime& dateAdded) {
+
+void Track::setLastPlayedDate(const QDateTime& playedDate) {
+    QMutexLocker lock(&m_qMutex);
+    if (playedDate == m_lastPlayed) {
```
There is a small utility function for setting track properties: `compareAndSet()`

ywwg replied:
done

ywwg committed:
last played: address notes

ywwg committed:
last played: make const extern

ywwg committed:
last played: fix naming

ywwg committed:
last played: persist and reuse QSqlQuery

Split lastplayed helpers into two separate objects, one to
create the table and the other to read from it.  They are used
by different parts of the code and need access to different things.

ywwg committed:
last played: check for not-found

ywwg committed:
last played: make sure query is finished

ywwg committed:
Merge branch 'master' into last-played

ywwg committed:
Merge branch 'master' into last-played

ywwg committed:
Merge branch 'master' into last-played

daschuer commented:
Also: https://bugs.launchpad.net/mixxx/+bug/1852925

ywwg committed:
Merge branch 'master' into last-played

ywwg committed:
Merge branch 'master' into last-played

ywwg committed:
Last Played: fix not reprocessing tracks from deleted playlists

ywwg commented:
Fixed an issue where deleting a playlist did not correctly update the tables.

I also did a drive-by change to the trackdeleted signal, which had a "position" parameter that was never used.

ywwg committed:
Last Played: minor cleanup

ywwg committed:
Last Played: line wrap

ywwg committed:
Merge branch 'master' into last-played

ywwg committed:
Merge branch 'master' into last-played

ywwg committed:
Merge branch 'master' into last-played

ywwg committed:
Fix bad merge

uklotzde commented on src/library/basesqltablemodel.cpp:
```diff
@@ -88,7 +88,7 @@ void BaseSqlTableModel::initSortColumnMapping() {
     m_columnIndexBySortColumnId
             [TrackModel::SortColumnId::SORTCOLUMN_LASTPLAYEDDATETIME] =
-                    fieldIndex(ColumnCache::COLUMN_LASTPLAYED_DATETIMEPLAYED);
+                    fieldIndex(ColumnCache::COLUMN_LIBRARYTABLE_DATETIMEPLAYED);
```
Commits in the history that do not even build are bad for git bisect. Redo and force push.

ywwg commented:
obsoleted by #3140
